### PR TITLE
[TVMSCRIPT] add more type support in script function parameter

### PR DESCRIPTION
**Pull request by Beya2019**
Now the function parameters decorated by @tvm.script.tir only support int32 and handle.
If we have such a demand: func parameters need to be int8 float32 and float16, we can only  write this
(if c is a float32 data)
```
@tvm.script.tir
def func(a: ty.handle, out:ty.handle) -> None:
    out_buf = tir.match_buffer(out, (1, 5), "float32")
    in_buf = tir.match_buffer(a, (1), "float32")
    with tir.block([1, 5], "init") as [vi, vj]:
        out_buf[vi, vj] = in_buf[0]

np_data = np.zeros(shape=(1, 5)).astype("float32")
score = 0.5
out_data = tvm.nd.array(np_data, ctx)
score_data = tvm.nd.array(np.array([score], dtype="float32"), ctx) 
f(score_data, out_data) 
```
Would it be better like this(Under the premise of defining ty.float32) 
```
@tvm.script.tir
def func(a:ty.float32, out:ty.handle) -> None:
    out_buf = tir.match_buffer(out, (1, 5), "float32")
    with tir.block([1, 5], "init") as [vi, vj]:
        out_buf[vi, vj] = a

np_data =np.zeros(shape=(1, 5)).astype("float32")
score = 0.5
out_data = tvm.nd.array(np_data, ctx)
f(score, out_data) 
```
In the future, if we use script to define more operators like the iou_threshold non_max_suppression and on_value/off_value in one_hot, this will simplify our coding IMO.

### CODE REVIEW

**Comment by tkonolige** [comment]
@tqchen Is this something we wish to support?

**Comment by tqchen** [comment]
Thanks @Beya2019 , can you add a regression testcase? Right now we do support passing in value arguments

**Comment by tkonolige** [comment]
Can you also add `int16`, `int64`, and `float64`?

**Comment by Beya2019** [comment]
> Thanks @Beya2019 , can you add a regression testcase? Right now we do support passing in value arguments

**Comment by Beya2019** [comment]
> > Thanks @Beya2019 , can you add a regression testcase? Right now we do support passing in value arguments

**Comment by Beya2019** [comment]
> Thanks @Beya2019 , can you add a regression testcase? Right now we do support passing in value arguments

Already added the get_valid_counts in test_tvmscript_ops.py
```
@tvm.script.tir
def get_valid_counts(
    data: ty.handle,
    valid_count: ty.handle,
    out: ty.handle,
    out_indices: ty.handle,
    score_threshold: ty.float32,
    id_index: ty.int32,
    score_index: ty.int32,
) -> None:
```

In this operator, the data type of score_threshold is ty.float32 which is our newly supported situation

**Comment by Beya2019** [comment]
> Can you also add `int16`, `int64`, and `float64`?

OK, Now support types as below：
```
int8 = ConcreteType("int8")
int16 = ConcreteType("int16")
int32 = ConcreteType("int32")
int64 = ConcreteType("int64")
float16 = ConcreteType("float16")
float32 = ConcreteType("float32")
float64 = ConcreteType("float64")
bool = ConcreteType("bool")
handle = ConcreteType("handle")
```

File: tests/python/unittest/test_tvmscript_ops.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    out_indices: ty.handle,
+    score_threshold: ty.float32,
+    id_index: ty.int32,
+    score_index: ty.int32,
+) -> None:
+
+    data_buf = tir.match_buffer(data, (1, 2500, 6), "float32")
+    valid_count_buf = tir.match_buffer(valid_count, (1,), "int32")
+    out_buf = tir.match_buffer(out, (1, 2500, 6), "float32")
+    out_indices_buf = tir.match_buffer(out_indices, (1, 2500), "int32")
```
Comment (jcf94) [inline tests/python/unittest/test_tvmscript_ops.py:22]: Thanks. I wonder what if the buffer type is not matched with the parameter? Suggest to also add a bad case test.

Comment (Beya2019) [reply tests/python/unittest/test_tvmscript_ops.py:22]: if the buffer type is not matched with the parameter,  an error will be reported as below When it is bulit,
`Check failed: (dtype.element_of() == n->dtype.element_of() && dtype.lanes() % n->dtype.lanes() == 0) is false: Cannot store float32 to buffer of float16` 

IMO, This error is not only reflected in the buffer type is not matched with the parameter， two buffer type which is not matched, the same problem will also occur. eg (a buffer is float32, out buffer is float16):

```
@tvm.script.tir
def func(a: ty.handle, out:ty.handle) -> None:
    out_buf = tir.match_buffer(out, (1, 5), "float16")
    in_buf = tir.match_buffer(a, (1), "float32")
    with tir.block([1, 5], "init") as [vi, vj]:
        out_buf[vi, vj] = in_buf[0]

np_data = np.zeros(shape=(1, 5)).astype("float16")
score = 0.5
out_data = tvm.nd.array(np_data, ctx)
score_data = tvm.nd.array(np.array([score], dtype="float32"), ctx) 
f(score_data, out_data) 

```

So when we write script functions, we must avoid such errors.

**Review by jcf94 — APPROVED** [review]

**Comment by jcf94** [comment]
Hi @Beya2019, let's fix these CI errors and this PR will be fine to merge.

**Comment by Beya2019** [comment]
> Hi @Beya2019, let's fix these CI errors and this PR will be fine to merge.

Already fix CI errors. Please help to merge it. Thanks very much.

**Comment by jcf94** [comment]
> > Hi @Beya2019, let's fix these CI errors and this PR will be fine to merge.
> 
> Already fix CI errors. Please help to merge it. Thanks very much.

Great! @Beya2019 Now merged.
Thanks! @tkonolige @tqchen

**Pull request merged**